### DISCUSSION TITLE
Ignore some categories for imputation

### DISCUSSION THREAD

**Discussion by MuzzJ** [amices/mice#640] [Q&A] [answered]
Dear MICE community,

I would like to ask for help and advice.
I have a dataset where one nominal variable has 4 values, all have some presence in the dataset. Some data were lost, but it is known that the missing values can be two of the four categories.

While trying to add constraints, I came across the discussion where Prof. Stef van Buuren recommended squeeze_vec() function here: https://github.com/amices/mice/discussions/507 

I tried the squeeze_vec() function and it didn't work with factor variables.
I tried to implement the following function then:

`
#Possible values are 1,2,3,4
#I want to impute 1 and 2.

mice.impute.School <- function(y, ry, x, ...) {
  selected_obs <- !ry | (y %in% c("1", "2"))
  y_miss <- mice.impute.polyreg(
    forcats::fct_drop(y[selected_obs]),
    ry[selected_obs],
    x[selected_obs, ], ...
  )
  return(y_miss)
}`

The results are a bit disappointing:
`#Initial data:
  1    2    3    4 <NA> 
  10   28  205  193  105 

#One imputation with that function in the methods and there are NAs
imp_2 <- mice(
  df, pred=predM, meth = meth, maxit = 1, m = 5, printFlag = TRUE, ridge=0.01
)
> table(imp_n1$School, useNA="always")

   1    2    3    4 <NA> 
  27   51  234  227    2

#Using map function reduced the number in category 3 and 4 and there are NAs
> map(1:5, \(x) complete(imp_2, x)) %>% 
+   map(\(d) d$School[is.na(df$School)]) %>% 
+   unlist() %>% table(useNA = "ifany")
.
   1    2    3    4 <NA> 
 129  126  134  126   10`

Here is the code:
`input.data.path <- paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/")
setwd(input.data.path)

#Loading the data
df = read_excel(path = '4unies_quantitative_no_sds_clean_noid.xlsx')

# Convert columns to factors using base R

cols<-c("gender","language","education","marital","exchProg","cctrain", "School") # Creates a vector of column names
df[cols]<-lapply(df[cols],as.factor)

library(mice)
set.seed(1234)

imp<-mice(df,maxit = 0,print=F) # a dry run.

predM<-imp$predictorMatrix # in the predictor matrix: columns are predictors, rows are predicted vars.

predM[,c("exchProgDue", "sf23", "sf24", "sf25", "sf26", 
         "sf27","sf28","sf29","sf30","sf31")]<-0 #  these vars aren't used as predictors for any vars.

predM

# Setting a method for predicting each variable.
meth<-imp$method

meth[c("as1","as2","as3",
       "cca1","cca2","cca3",
       "sf1","sf2","sf3","sf4","sf5","sf6","sf7","sf8","sf9","sf10",
       "sf11","sf12","sf13","sf14","sf15","sf16","sf17","sf18","sf19","sf20",
       "sf21","sf22","sf23","sf24","sf25","sf26","sf27","sf28","sf29","sf30",
       "sf31","sf32","sf33","sf34","sf35","sf36",
       "soc1","soc2","soc3","soc4","soc5","soc6","soc7","soc8","soc9","soc10",
       "soc11","soc12","soc13",
       "es1","es2","es3","es4","es5","es6","es7","es8","es9","es10",
       "ccm1","ccm2","ccm3","ccm4","ccm5",
       "a1","a2","a3","a4",
       "pss1","pss2", "pss3", "pss4","pss5","pss6", "pss7", "pss8", "pss9", "pss10",
       "age")] <- "pmm"        
meth[c("gender","ses","language","education","marital",
       "exchProg","exchProgDue","cctrain","cctraindue")]<-""
# Assign the custom method directly by column index # we will impute only 1, 2

mice.impute.School <- function(y, ry, x, ...) {
  selected_obs <- !ry | (y %in% c("1", "2"))
  y_miss <- mice.impute.polyreg(
    forcats::fct_drop(y[selected_obs]),
    ry[selected_obs],
    x[selected_obs, ], ...
  )
  return(y_miss)
}

meth["School"] <- "School"

imp_2 <- mice(
  df, pred=predM, meth = meth, maxit = 1, m = 5, printFlag = TRUE, ridge=0.01
)

imp_n1<-complete(imp_2,1)
table(df$School, useNA="always")
table(imp_n1$School, useNA="always")

map(1:5, \(x) complete(imp_2, x)) %>% 
  map(\(d) d$School[is.na(df$School)]) %>% 
  unlist() %>% table(useNA = "ifany")
####`

**Comment by stefvanbuuren** [comment, marked as answer]
>>> I have a dataset where one nominal variable has 4 values, all have some presence in the dataset. Some data were lost, but it is known that the missing values can be two of the four categories.

I suggest you try the `exclude` argument from `mice.impute.pmm()`. It specifies a vector of dependent values to exclude from the imputation model and the collection of donor values. Excluding the two impossible categories would then always impute one of the other two.

**Reply by MuzzJ** [reply]
Thank you very much!

I finetuned it and now all works:
```
# Use blots to exclude different values per column
# Create blots object
blots <- make.blots(df)
# Exclude ml 3 through 4 from School donor pool
blots$School$exclude <- c("3","4")


imp <- mice(df, method = meth, pred =predM, maxit = 1,ridge = 0.01, 
            m = 1,  print = TRUE, blots = blots)
```

The only downside is that it leaves a couple on N/A's, I am trying to figure out why..

update:
I have figured how to fix it, I didn't specify the methods (I have several demographic variables, they are all filled in, so I tried to exclude them from being imputed. When I changed all methods to "pmm" it worked perfectly fine):

```
imp <- mice(df, m = 1, method = "pmm", pred =predM,
           blots = blots,  print = TRUE)
```
